src/main/java/duke/Storage.java

```diff
@@ -0,0 +1,141 @@
+            }
+        } catch (FileNotFoundException e) {
+            System.out.println("An error occurred.");
+            e.printStackTrace();
+        } finally {
+            return list;
+        }
+    }
+
+    /**
+     * Read task from the filePath.
+     *
+     * @param inputFile File which the tasks are read from.
+     * @return An ArrayList of Tasks which the tasks read are added into.
+     */
+    private ArrayList<Task> readTasks(File inputFile) {
+        ArrayList<Task> list = new ArrayList<>();
+        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("dd MMM yyyy");
+        try {
+            Scanner s = new Scanner(inputFile);
+            BufferedReader reader = new BufferedReader(new FileReader(inputFile));
+            while (s.hasNextLine()) {
+                String line = s.nextLine();
+                String type = line.substring(1, 2);
+                String mark = line.substring(4, 5);
+
+                if (type.equals("T")) {
+                    String description = line.substring(7);
+                    Todo t = new Todo(description);
+                    if (mark.equals("X")) {
+                        t.markAsDone();
+                    }
+                    list.add(t);
+                } else if (type.equals("D")) {
+                    String temp = line.substring(7);
+                    String[] result = temp.split("\\(by: ");
+                    String description = result[0].trim();
+                    String by = result[1].substring(0, result[1].length() - 1);
+                    LocalDate date = LocalDate.parse(by.substring(0, 11), dateTimeFormatter);
+                    LocalTime time = LocalTime.parse(by.substring(12, 17));
+                    Deadline t = new Deadline(description, date, time);
+                    if (mark.equals("X")) {
+                        t.markAsDone();
+                    }
+                    list.add(t);
+                } else if (type.equals("E")) {
+                    String temp = line.substring(7);
+                    String[] temp2 = temp.split("\\(from: ");
+                    String description = temp2[0];
+                    String result = temp2[1];
+                    LocalDate startDate = LocalDate.parse(result.substring(0, 11), dateTimeFormatter);
+                    LocalTime startTime = LocalTime.parse(result.substring(12, 17));
+                    LocalDate endDate = LocalDate.parse(result.substring(24, 35), dateTimeFormatter);
+                    LocalTime endTime = LocalTime.parse(result.substring(36, 41));
+                    Event t = new Event(description, startDate, startTime, endDate, endTime);
+                    if (mark.equals("X")) {
+                        t.markAsDone();
+                    }
+                    list.add(t);
+                }
+            }
+        } catch (FileNotFoundException e) {
+            System.out.println("An error occurred.");
```

Comment:
Maybe can consider customizing the error messages for different types of errors to enhance the user experience, instead of solely relying on printStackTrace(). Besides that, I am not sure if the string parsing approach used here is a bit of hard coding, which could lead to maintenance issues if the structure of the input data changes. For example, if 'event' is modified to another word, all the substrings with specific indices need to be adjusted accordingly. 

One alternative approach is to use a delimiter, such as "|" to separate different fields in the input data. This would allow for easier parsing using methods like split() rather than relying on substrings with specific indices.